### PR TITLE
node drain timeout via labels

### PR DESCRIPTION
**Pull request by grosser**
*Issue #, if available:*
https://github.com/aws/karpenter/issues/2481

*Description of changes:*
allow setting `karpenter.sh/drain-timeout: "1h"` label to force-drain nodes that take too long

alternatively this could be done with a new field in the provisioner spec, but we did not see a way of accessing that from the `Reconcile`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

/cc @Timothy-Dougherty @yizhang-zen

### CODE REVIEW

**Comment by yizhang-zen** [comment]
hi @jonathan-innis , does this look good? Is there anything need to be changed?

**Comment by jonathan-innis** [comment]
Hey @grosser and @yizhang-zen, I apologize for being slow on this one, got caught up with some other things. @dewjam is currently working on an RFC that should work towards a better solution for nodes that are stuck in a draining state for an indefinite amount of time (or forever potentially!).

At a high-level, we are pretty cautious of implementing a drain timeout or force drain functionality, primarily because it goes against the Kubernetes eviction API and we are trying to keep in line with what Kubernetes accepts as feasible for node pod eviction. I'll let @dewjam comment more on his thoughts around it though since he is owning this design.

**Comment by jonathan-innis** [comment]
I'd like to see further discussion on this one/the RFC to get a feel for what everyone in the community feels is the best way to solve this problem of hung nodes.

**Comment by grosser** [comment]
Other approaches:
- "make this a flag in the provisioner". This has the upside of not being a 1-off label, but the downside of not being able to force-drain a node via `kubectl label`
- "make this an annotation" blocked on that field not existing

... so I'd want to know:
- did we miss any approach ?
- is anyone from aws side planning to work on this and when ?
- can this be merged as-is or should we rework it in some way ?

I don't think there will be any meaningful community discussion since the original issue https://github.com/aws/karpenter/issues/2481 had no discussion, so prefer to pick any solution you deem good enough and get that implemented (either by us with some guidance or by aws).

**Comment by ellistarn** [comment]
tl;dr, we've been discussing this internally, but have had other priorities overrule this. I think we should probably bring https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown into the discussion, as well.

**Comment by grosser** [comment]
If someone uses that they would not opt-in the the termination here,
but afaik we still need a termination timeout here.
- Is this PR good enough to merge and maybe not document or mark as alpha ?
... or are you thinking the feature might not be needed at all ?
... or are you thinking it's not the right approach to implement?

**Comment by bwagner5** [comment]
I agree with Ellis on looking at the Graceful Node Shutdown K8s feature to guide this or see if it can fill this gap as is:

```
shutdownGracePeriod:
    Specifies the total duration that the node should delay the shutdown by. 
    This is the total grace period for pod termination for both regular and [critical pods](https://kubernetes.io/docs/tasks/administer-cluster/guaranteed-scheduling-critical-addon-pods/#marking-pod-as-critical).
shutdownGracePeriodCriticalPods:
    Specifies the duration used to terminate [critical pods](https://kubernetes.io/docs/tasks/administer-cluster/guaranteed-scheduling-critical-addon-pods/#marking-pod-as-critical) during a node shutdown. 
    This value should be less than shutdownGracePeriod.
```

**Comment by grosser** [comment]
we don't trigger node-shutdown from karpenter until the drain is complete, so this feature would never be triggered

**Comment by grosser** [comment]
... unless you are suggesting that karpenter does not draining at all and just terminates the node, then let's graceful shutdown handle things
that would work too and is most likely the way we are going since otherwise we end up with 2 draining systems

**Comment by ellistarn** [comment]
Generally, we're wary of violating kubernetes API specification. In this case, we have PDBs and GracefulTerminationPeriods in play.

Also, we just changed our behavior where naked pods could block node shutdown. We're not aware of other cases where draining can get blocked, other than misconfigured PDBs. In that case we recommend configuring the PDBs correctly, rather than violating kubernetes API constraints.

Are there concrete cases we're missing where you want to force deletion?

**Comment by grosser** [comment]
Our blockers are (now fixed) naked pods and misconfigured pdbs.
I'm not suggesting that timeout should be a default behavior, but I think they would be a nice option for test-clusters or staging environments where broken apps/pdbs can exist.

**Comment by ellistarn** [comment]
Gotcha. At this point, I'm not supportive of this feature for the misconfigured PDB case. Those misconfigured pdbs will break assumptions of other aspects of kubernetes (e.g. rollout), so I think you're best off relaxing the PDBs in dev environments. Is that okay with you?

**Comment by grosser** [comment]
Our workaround could be to auto-delete PDBs that are misconfigured for too long (or making them not be invalid by allowing 999 violations etc).

The other approach I wanted to try is to allow disabling draining in karpenter since we already have a tool that does draining and it would be nice to have all our logic consolidated, but that will be a new PR.

**Comment by ellistarn** [comment]
> The other approach I wanted to try is to allow disabling draining in karpenter 

Generally, we're wary of adding complexity/api surface for one off niche use cases. There are real tradeoffs here.

Why not just let karpenter drain, and your tool also drain to unstick the extra pods?

**Comment by ellistarn** [comment]
Btw, appreciate all of your effort and ideation across these PRs. Feel free to reach out on slack at anytime @ellistarn, and I'm happy to brainstorm ideas in real time.

**Comment by grosser** [comment]
I'm trying to keep discussion on github so everybody can find them and to be async :)

The idea for disabling is mostly around:
- unified status reporting (we can reuse our dashboards and monitors for all drain activity)
- using the same safeguard logic (for example we don't remove nodes when a neighboring group is already in trouble, or make sure to only drain x% of a group)
... it should be easy to do with a fork too, I understand that it's might not be valuable to karpenters userbase.

**Comment by ellistarn** [comment]
Do you still want karpenter to make termination decisions if you're handling cordon/drain?

**Comment by grosser** [comment]
no, we'd disable the whole drain/termination ... just have it "delete the node" so it it's terminating.

**Comment by njtran** [comment]
Hey @grosser, we appreciate the discussion you've had with us on this end. Similar to what @ellistarn and @jonathan-innis have said above, we're generally wary about introducing more complexity on the API surface. We are working on an RFC that should work towards a better solution for nodes that are stuck in a draining state for an indefinite amount of time, but IIUC, this doesn't fix your case for the misconfigured PDBs.

I'm going to close this PR as I don't think we'll be able to support this one-off case. A node drain timeout on misconfigured PDBs could introduce unexpected disruptions on workloads, and is counterintuitive to the use of PDBs. If you have any more comments on this, feel free to reach out to us on the kubernetes/karpenter slack channel. 

If you want better controls on how to detect misconfigured PDBs, [this is how](https://github.com/aws/karpenter-core/blob/main/pkg/controllers/termination/eviction.go#L93-L116) Karpenter responds to eviction API failures, where a misconfigured PDB is a 500. If you're able to programmatically find occurrences of the misconfigured PDB logs in your controller logs, you might be able to detect this case quicker.
